staging/src/k8s.io/apimachinery/pkg/api/resource/amount_test.go

```diff
@@ -88,6 +88,45 @@ func TestInt64AmountAdd(t *testing.T) {
 		}
 	}
 }
+
+func TestInt64AmountMul(t *testing.T) {
+	for _, test := range []struct {
+		a  int64Amount
+		b  int64
+		c  int64Amount
+		ok bool
+	}{
+		{int64Amount{value: 100, scale: 1}, 1000, int64Amount{value: 100000, scale: 1}, true},
+		{int64Amount{value: 100, scale: -1}, 1000, int64Amount{value: 100000, scale: -1}, true},
+		{int64Amount{value: 1, scale: 100}, 10, int64Amount{value: 1, scale: 100}, false},
+		{int64Amount{value: 1, scale: -100}, 10, int64Amount{value: 1, scale: -100}, false},
+		{int64Amount{value: -5, scale: 2}, 500, int64Amount{value: -2500, scale: 2}, true},
+		{int64Amount{value: -5, scale: -2}, 500, int64Amount{value: -2500, scale: -2}, true},
+
+		{int64Amount{value: mostPositive, scale: -1}, 10, int64Amount{value: mostPositive, scale: -1}, false},
+		{int64Amount{value: mostPositive, scale: -1}, 0, int64Amount{value: 0, scale: 0}, true},
+		{int64Amount{value: mostPositive / 10, scale: 1}, 10, int64Amount{value: mostPositive / 10, scale: 1}, false},
```

Comment:
Good test case, but why does this one fail?  `(MAX/10) * 10` should result in `MAX`, no?

Reply:
I think it should fail because the test case indicates `(MAX/10) * 10^1 * 10` where `10^1` means `{scale :1}`.

Reply:
I don't understand - 10^1 is 10.  X/10*10 = X - how does this fail?

Reply:
oh, man.  I misread.  Let me run thru it all again :)

Reply:
OK, so I see why this fails now, but I don't love it as an API.  It's probably not how I would have laid out the API.  

`int64Amount{value: mostPositive, scale: 2}` is a valid value, but it can't be converted to int64.

```
func TestFooBar(t *testing.T) {
    x := int64Amount{value: mostPositive, scale: 2}
    d := x.AsDec()
    i, err := x.AsInt64()
    t.Errorf("\n  x: %v\n  d: %v\n  i: %v (%v)\nmax: %v", x, d, i, err, math.MaxInt64)                                                                                                                
}
```

gives:

```
          x: {9223372036854775807 2}
          d: 922337203685477580700
          i: -100 (false)
        max: 9223372036854775807
```

The docs for `type int64Amount` don't say anything about this, but the fact that `AsInt64()` returns a bool suggests that it's understood that it can hold values that are larger than int64.

`Add()` says it fails " if overflow or underflow would result" but not loss of precision, though I guess we can infer that.  So `Mul()` is at least consistent.

This code is filled with hand-optimizations that I bet we don't need any more.

This comment is for the record, no action needed,